### PR TITLE
feat: Hyperopt schema v0, part 1: Move output feature metrics from feature classes to feature configs.

### PR DESCRIPTION
**Pull request by ksbrar**
This is part 1 of the broken up hyperopt schema PR (https://github.com/ludwig-ai/ludwig/pull/2346).

It moves the definition of a feature's metric function definitions from that feature's class to its config. This seems to make more sense overall and it is necessary to populate the JSON schema for the output feature metric section of the hyperopt config.

### CODE REVIEW

File: ludwig/features/binary_feature.py [hunk 1]
```diff
@@ -252,8 +239,6 @@ def create_preproc_module(metadata: Dict[str, Any]) -> torch.nn.Module:
 
 
 class BinaryOutputFeature(BinaryFeatureMixin, OutputFeature):
-    metric_functions = {LOSS: None, ACCURACY: None, ROC_AUC: None}
-
```
Comment (connor-mccorm) [inline ludwig/features/binary_feature.py:256]: Seems like you are missing a call to `BinaryOutputFeatureConfig.get_output_metric_functions()` here right?

**Review by connor-mccorm — APPROVED** [review]
LTGM except it seems like you still need to add the `get_output_metric_functions()` call to the binary feature. Great job mate!

**Review by dantreiman — APPROVED** [review]
Had a few code style edits, let me know what you think of those.  Otherwise looks good so approved.

File: ludwig/features/binary_feature.py (hunk 1 above)
Comment (ksbrar) [reply ludwig/features/binary_feature.py:256]: fixed

File: ludwig/features/category_feature.py [hunk 2]
```diff
@@ -240,7 +237,7 @@ def create_preproc_module(metadata: Dict[str, Any]) -> torch.nn.Module:
 
 
 class CategoryOutputFeature(CategoryFeatureMixin, OutputFeature):
-    metric_functions = {LOSS: None, ACCURACY: None, HITS_AT_K: None}
+    metric_functions = CategoryOutputFeatureConfig.get_output_metric_functions()
```
[the remark this answers is not in the document]
Comment (ksbrar) [reply ludwig/features/category_feature.py:240]: I like this much better!

[the remark this answers is not in the document]
Comment (justinxzhao) [reply ludwig/features/category_feature.py:240]: This would be cleaner, but there might be some complexity with how `OutputFeature` base class shadows the class attribute, with `def _setup_metrics(self)`.

IMO, probably worth refactoring/cleaning this up separately.

**Review by justinxzhao — APPROVED** [review]
Hi Kabir, happy to go with this PR as is. There's some additional cleanup that @dantreiman suggested, but this PR as is is still cleaner than what we have today, and we could defer the additional refactoring to later, as to continue with checking in the rest of the hyperopt schema changes, WDYT?

**Comment by ksbrar** [comment]
@justinxzhao Sounds good, I will proceed with the other changes first and put the set change in a separate refactor!
